src/parsing/module.rs

```diff
@@ -0,0 +1,112 @@
+use std::collections::HashSet;
+use std::path::PathBuf;
+
+use crate::core::config::ModuleConfig;
+use crate::core::module::ModuleTree;
+use petgraph::algo::kosaraju_scc;
+use petgraph::graphmap::DiGraphMap;
+
+use super::error::ModuleTreeError;
+use super::py_ast::parse_interface_members;
+
+pub fn find_duplicate_modules(modules: &[ModuleConfig]) -> Vec<&String> {
+    let mut duplicate_module_paths = Vec::new();
+    let mut seen = HashSet::new();
+
+    for module in modules {
+        if seen.contains(&module.path) {
+            duplicate_module_paths.push(&module.path);
+        } else {
+            seen.insert(&module.path);
+        }
+    }
+
+    duplicate_module_paths
+}
+
+pub fn find_modules_with_cycles(modules: &[ModuleConfig]) -> Vec<&String> {
+    let mut graph = DiGraphMap::new();
+
+    // Add nodes
+    for module in modules {
+        graph.add_node(&module.path);
+    }
+
+    // Add dependency edges
+    for module in modules {
+        for dependency in &module.depends_on {
+            graph.add_edge(&module.path, &dependency.path, None::<()>);
+        }
+    }
+
+    // Find strongly connected components (SCCs)
+    let sccs = kosaraju_scc(&graph);
+
+    // Filter SCCs to find cycles
+    let mut modules_with_cycles = Vec::new();
+    for scc in sccs {
+        if scc.len() > 1 {
+            modules_with_cycles.extend(scc);
+        }
+    }
```

Comment:
Very nice - do you have a feel for the runtime cost of this?

Reply:
kosaraju runs in `O(V+E)` (#Vertices + #Edges). The python version finds cycle using DFS starting on every node, and checks the cycle if the node exists in cycle which should be `O(VE(V+E))` loosely. Yet to delete the python implementation.